[WASM] Fix the assertion in ExternRefTable and rename it to ExternOrAnyRefTable

hyjorc1 opened the pull request:
#### aea780d518debdbdf8ec34a939ea425a82f2addb
<pre>
[WASM] Fix the assertion in ExternRefTable and rename it to ExternOrAnyRefTable
<a href="https://bugs.webkit.org/show_bug.cgi?id=279390">https://bugs.webkit.org/show_bug.cgi?id=279390</a>
<a href="https://rdar.apple.com/135214204">rdar://135214204</a>

Reviewed by David Degazio.

With the introduction of WasmGC, ExternRefTable should also accept
`nullexternref` type (aka null). This patch fixes that by adding
additional subtype check against to `externref` [2] since it&apos;s not a
subtype of `anyref` [1]. As a wasm table can contain any reference types [3],
this patch also rename the ExternRefTable to ExternOrAnyRefTable.

[1] <a href="https://webassembly.github.io/gc/core/syntax/types.html#heap-types">https://webassembly.github.io/gc/core/syntax/types.html#heap-types</a>
[2] <a href="https://webassembly.github.io/gc/core/valid/matching.html#heap-types">https://webassembly.github.io/gc/core/valid/matching.html#heap-types</a>
[3] <a href="https://webassembly.github.io/gc/core/valid/types.html#table-types">https://webassembly.github.io/gc/core/valid/types.html#table-types</a>

* JSTests/stress/table-with-noexternref-wasm-type.js: Added.
(catch):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::Table::visitDerived):
(JSC::Wasm::Table::tryCreate):
(JSC::Wasm::Table::grow):
(JSC::Wasm::Table::visitAggregateImpl):
(JSC::Wasm::ExternOrAnyRefTable::ExternOrAnyRefTable):
(JSC::Wasm::ExternOrAnyRefTable::clear):
(JSC::Wasm::ExternOrAnyRefTable::set):
(JSC::Wasm::ExternRefTable::ExternRefTable): Deleted.
(JSC::Wasm::ExternRefTable::clear): Deleted.
(JSC::Wasm::ExternRefTable::set): Deleted.
* Source/JavaScriptCore/wasm/WasmTable.h:

Canonical link: <a href="https://commits.webkit.org/283472@main">https://commits.webkit.org/283472@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/902252b7bfe79d31172ec8c2f9cc5e611b628a3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66292 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45666 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18911 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70324 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16902 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53464 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17184 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53178 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11784 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69359 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42111 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57389 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33819 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38781 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14770 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15778 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/59406 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60677 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15114 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72027 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/65536 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10247 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14504 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60499 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10279 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57453 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60802 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8459 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2082 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/87303 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10060 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41474 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15350 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42551 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43732 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42293 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddegazio commented on Source/JavaScriptCore/wasm/WasmTable.cpp:
```diff
@@ -38,14 +38,14 @@
 template<typename Visitor> constexpr decltype(auto) Table::visitDerived(Visitor&& visitor)
 {
     switch (type()) {
     case TableElementType::Externref:
```
It's kind of rough naming, but we should probably change this to `TableElementType::ExternOrAnyref` too.

hyjorc1 replied:
It seems the spec https://webassembly.github.io/gc/js-api/#tables prefers
```
enum TableKind {
  "externref",
  "anyfunc",
};
```
Maybe we can leave it for now.

ddegazio reviewed (approved):
r=me with comment.

hyjorc1 committed:
[WASM] Fix the assertion in ExternRefTable and rename it to ExternOrAnyRefTable
https://bugs.webkit.org/show_bug.cgi?id=279390
rdar://135214204

Reviewed by David Degazio.

With the introduction of WasmGC, ExternRefTable should also accept
`nullexternref` type (aka null). This patch fixes that by adding
additional subtype check against to `externref` [2] since it's not a
subtype of `anyref` [1]. As a wasm table can contain any reference types [3],
this patch also rename the ExternRefTable to ExternOrAnyRefTable.

[1] https://webassembly.github.io/gc/core/syntax/types.html#heap-types
[2] https://webassembly.github.io/gc/core/valid/matching.html#heap-types
[3] https://webassembly.github.io/gc/core/valid/types.html#table-types

* JSTests/stress/table-with-noexternref-wasm-type.js: Added.
(catch):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::Table::visitDerived):
(JSC::Wasm::Table::tryCreate):
(JSC::Wasm::Table::grow):
(JSC::Wasm::Table::visitAggregateImpl):
(JSC::Wasm::ExternOrAnyRefTable::ExternOrAnyRefTable):
(JSC::Wasm::ExternOrAnyRefTable::clear):
(JSC::Wasm::ExternOrAnyRefTable::set):
(JSC::Wasm::ExternRefTable::ExternRefTable): Deleted.
(JSC::Wasm::ExternRefTable::clear): Deleted.
(JSC::Wasm::ExternRefTable::set): Deleted.
* Source/JavaScriptCore/wasm/WasmTable.h:

Canonical link: https://commits.webkit.org/283472@main
